### DISCUSSION TITLE
mdast-util-remove-wiki-links and mdast-util-wiki-links

### DISCUSSION THREAD

**Discussion by aegatlin** [syntax-tree/unist#81] [Ideas] [answered]
I have seen https://github.com/landakram/remark-wiki-link and https://github.com/flowershow/remark-wiki-link-plus. I am still understanding their codebases, so apologize if I'm confused in this post.

I was thinking there is a way to turn a `mdast-paragraph` into `[mdast-paragraph, mdast-link, mdast-paragraph]`.

```
In the [[middle]] of the sentence. => ['In the ', '[middle](/middle)', ' of the sentence.']
```

I'm imagining a simple mdast util that will do this. As well as one that will remove them, too.

```
In the [[middle]] of the sentence. => In the middle of the sentence.
```

I naively began parsing a mdast paragraph node for its `p.children.map({ value } => ...)` before realizing that was probably naive in some way because I've never contributed to this community before. Wound up here after reading: https://github.com/syntax-tree/mdast#contribute

Over the coming weeks I will hope to drive out some solutions along these lines, so any feedback is much appreciated. I'm sure yall are busy, so thank you for taking the time to read. <3

(PS, I love this ecosystem. You are doing a great job :D <3 I'm a very happy user!)

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @aegatlin 👋 

For
> I'm imagining a simple mdast util that will do this. As well as one that will remove them, too.
> ```
> In the [[middle]] of the sentence. => In the middle of the sentence.
> ```

I'd recommend using an existing wikilink parser (like https://github.com/landakram/remark-wiki-link).
Then creating an AST plugin similar to https://github.com/remarkjs/remark-unwrap-images to unwrap the text inside wiki links.
https://unifiedjs.com/learn/ introduces working with ASTs and creating plugins.

**Reply by aegatlin** [reply]
Thank you for sharing the `remark-unwrap-images`, that is a helpful guide.

[remark-wiki-link](https://github.com/landakram/remark-wiki-link#usage) doesn't create a [mdast Link](https://github.com/syntax-tree/mdast#link), instead it puts many keys in the `data` field. I don't understand how this gets successfully transformed to html, but also, I naively want a mdast util that "simply" creates a mdast Link. Am I missing something? Perhaps I can reach out to the author and ask why they avoided using an mdast Link.

## PS / You can ignore the below! XD

I think I have enough to move forward, but in an overabundance of caution will at least provide context here (feel free to ignore this part):

I want to iterate through a collection of wiki docs (private, via Obsidian). If they are tagged `public: true`, I add them to a preliminary whitelist. Then, I iterate again, this time transforming markdown to html. At this juncture I will check if the wiki-link is public. If it is not I will "remove the wiki link", and if it is public I will "replace the wiki link with a normal link". This will be in the `mdast -> mdast` phase, prior to transforming into rehype.

My strategy would be to build a remark plugin that is a wiki-link handler much like `remark-wiki-link` with what I think is enough difference to warrant a new attempt:

1. Receive a `filter` function that returns either the link to use or falsy. This avoids passing a whitelist in to the utility. It's essentially a `getSlugFromName` function that can return falsy too. 
2. If falsy, remove the wiki-link.

My gut is telling me this is a noisy utility, and therefore the wrong abstraction, but until I come up with something better, it can at least get my feet wet.

I'd love any feedback, but understand if you didn't read XD
Thank you for the feedback thus far, ❤️

**Reply by wooorm** [reply]
>  successfully transformed to html

https://github.com/syntax-tree/mdast-util-to-hast#fields-on-nodes

>  I naively want a mdast util that "simply" creates a mdast Link.

From the rest of your comments: no, you do not want a “clean” `Link` at all. Maybe you’d want it at the end. But from all your comments, it sounds a lot like you want all the metadata about aliases, permalinks, exists, etc, so you can transform link how you want.

> Perhaps I can reach out to the author and ask why they avoided using an mdast Link.

This seems to be answered by your further points, e.g.,:

> If they are tagged public: true, I add them to a …

**Reply by aegatlin** [reply]
Thank you for sharing the `#fields-on-nodes` section! That was information I had not found on my own, and it added a lot of clarity for me. 😄 🙏 

I understand why you think I don't _really_ want an mdast Link based on my current goals, @wooorm . I will reach out to the current maintainer of remark-wiki-links to discuss further. However, I want to make the case that I indeed do "merely" want an `mdast -> mdast` plugin. Curious what you all think about it.

### Long-winded ignorable bit below 😄 

Wiki links have a lot of downsides, most notably, not to spec vis a vis commonmark. At the same time, many local text editors will open normal markdown links that point to a relative file path (can confirm for Obsidian and VSCode). To elaborate, VSCode does not know what to do with `[[Other Doc]]`, but it can open `[Other Doc](./some/path/to/other-doc.md)`. Obsidian can open both. This creates an opportunity (whether silly, serious, or otherwise) to remove wiki links from a local dir of markdown, thereby making it spec-compliant markdown usable across text editors.

Furthermore, if such a utility were to be used across workflows (`md -> md` and `md -> hast`, but also any others), and if you don't want the bells and whistles of a bespoke solution (such as the current wiki-links one, given that it has `node.data` bespoke to html-transforms), then, the only flexibility required in the resultant markdown is the URI of the link.

For these reasons I am enticed by a `md -> md` utility that handles URI's flexibly. This is why I want a mdast util, and this is why I want a `getUri` / `filter` function that can react to the environment in which it is called. But, huge caveat, I am naive in this space, and so want to hear the opinions of you wiser folk before trying to do this.

To that end, thank you so much for all the feedback thus far, and for creating these awesome tools <3

**Reply by aegatlin** [reply]
I have reached out the author in an issue here: https://github.com/landakram/remark-wiki-link/issues/23

**Reply by wooorm** [reply]
Have you thought about making your own plugin, based on `remark-wiki-links`?
I feels to me like you’re asking about that, but perhaps wondering how to do it?

**Reply by aegatlin** [reply]
I think I wanted a gut check where if yall said "don't build it, it's stupid", I'd move on from the idea. Since that didn't happen, I will go ahead and build a PoC. I have learned alot about how to get started building plugins from this thread, so hopefully I can smush some code together and then we could move forward from there 😆 

❤️

**Reply by aegatlin** [reply]
@wooorm @ChristianMurphy here is a proof of concept: https://github.com/aegatlin/remark-wiki-links-to-links

**Reply by wooorm** [reply]
Why not first use the existing plugin, and then change things?
This looks very minimal, pretty far still from what you want to achieve?

**Reply by aegatlin** [reply]
Okay, good question, here's my reasoning for not wanting to use https://github.com/landakram/remark-wiki-link (which is otherwise a great library serving its use case well):

1. It's _for_ html. It returns the following:

```
{
    value: 'Test Page',
    data: {
        alias: 'Test Page',
        permalink: 'test_page',
        exists: false,
        hName: 'a',
        hProperties: {
            className: 'internal new',
            href: '#/page/test_page'
        },
        hChildren: [{
            type: 'text',
            value: 'Test Page'
        }]
    }
}
```

(Based on its test suite, it must also return `type: 'wikiLink'`, but maybe it gets removed at some point.)

I don't want to presume html. I want open ended use-cases to be supported. E.g., I want people to be able to download md or plaintext, or include this in a markdown-formatting tool so they can write md without wiki-links. All these use cases have no desire for for any `h*` fields because those are hast-specific. I also don't want the other fields, alias, permalink, and exists. Those are useful fields when building a web-view for a wiki, but even in my own html use-case I'm not doing that. I'm only sourcing some content from Obsidian in a CMS-like way.

2. My desired use case is simpler than the current tool's use case. There will eventually be some replicated behavior around managing aliases (if move beyond the PoC) and the `toUri` func will be similar to the `pathResolver` func, but other than that, it would be unnecessarily complicated to try to leverage the functionality of the current tool to achieve my goals.

3. These could lead to (what I think could be) useful remark-only utilities. Implicit in the PoC are utilities other people might want for other use cases, like `mdast-util-remove-wiki-links`, perhaps. Smaller, simple tooling devoted to wiki-style-link manipulation could lead to more use cases.

Curious what you think about some of these points 😄 . And, thanks for taking the time to work with me on this. I really appreciate it ❤️

**Reply by aegatlin** [reply]
Oh, and one last thing

> This looks very minimal, pretty far still from what you want to achieve?

This is actually pretty much all I want to achieve.

**Reply by wooorm** [reply]
> I don't want to presume html.

Sure, but you can take the nodes that plugin generates, and map them to the nodes you want.
That’s much better that regexing text in a tree.
You will run into tons of bugs by regexing constructs like this!

> My desired use case is simpler than the current tool's use case

The hardest part is parsing. The complexity in that plugin is the parsing.
After that, there are ASTs, and transforming those ASTs is comparatively much simpler.

> useful remark-only utilities.

Totes, agreed. That’s your goal. I think there’s a better way to reach that goal than regexing though!

**Reply by aegatlin** [reply]
I think this is what I needed to hear. Thank you for the feedback. I think I have some more research to do! 🙏

**Reply by aegatlin** [reply]
Hello, here is the updated utility PoC: https://github.com/aegatlin/remark-wiki-links-to-links

Notably, the [example.js file](https://github.com/aegatlin/remark-wiki-links-to-links/blob/main/example.js) where I printed the mdast pre and post transform.

I imagine there are still good objections coming, so I want to make sure that I give you the 👍  to continuing to critiquing what I've done, and what I'm thinking. So, please critique away! Here are some of my thoughts:

1. This seems good enough conceptually?
1. I still don't love the API options, `{ toUri }`, but I think it's simplicity is an improvement for use cases similar to mine. 
1. It would be nice to eventually replace the underlying micromark extension ([micromark-extenstion-wiki-link](https://github.com/landakram/micromark-extension-wiki-link)) with something that doesn't presume an eventual html target. It could perhaps be something as "simple" as `{ type: 'wikiLink', value: [insert-value-here], data: { alias: [insert-alias-here] } }`. This could be a transparent swap to users of this lib.
1. Such an extension could trickle up to a mdast extension. Essentially mimicking the original extensions all the way up, minus the presumption of html at the end. This could be a transparent swap to users of this lib.
1. In the lib itself I feel like I should "just" use unist-util-visit to replace the node, but couldn't figure out how to do that from the docs. It uses SKIP somehow, but was too confused. If there's a good example of using visit instead of map to do simple replacements, I could update the docs for visit.
1. Alternatively, perhaps a `unist-util-replace` with a similar api to `unist-util-visit` could be a useful addition to the community?

I hope I don't sound too presumptuous about what I've done. This is all thanks to yalls great feedback, and I hope to continue receiving it. Thank you for everything thus far 🙏 ❤️

**Reply by wooorm** [reply]
1. yes
2. agreed, but I’ll leave that up to you. I don’t know enough about your use case
3. I disagree, `micromark-extension-wiki-link`, like any micromark extension, does not describe the AST. It would not need changing for yo to reach your goals. It would be somewhere in `mdast-util-wiki-link` instead
4. sure but depends on 3
5. https://unifiedjs.com/learn/recipe/remove-node/#replacing-a-node-with-its-children.
   I’d **extremely strongly** suggest using `visit` for everything. Do not use `filter`/`map`
6. I don’t know what you are thinking of `unist-util-replace` to perform

**Reply by aegatlin** [reply]
Hey all, thank you for all the feedback. I ended up reworking some of my stuff based on all this amazing feedback, and currently I have [remark-obsidian-link](https://github.com/aegatlin/remark-obsidian-link). 

I'm using it on my website to parse obsidian. My website is private but I have some utility functions from it in another repo call [metamark](https://github.com/aegatlin/metamark), within which I build a `toLink` function from a `pageAllowSet` provided by the user of metamark (code [here](https://github.com/aegatlin/metamark/blob/main/lib/toLinkBuilder.ts#L12)).

It still has a `toLink` function, which I'm still not in love with, but still can't think of a better way. I like the runtime-nature of `toLink`, as well as the minimal, custom typings of `WikiLink` and `Link`. But, I also know these things might be too bespoke to be useful to the community writ large. I hope to clarify my thinking further as I work with these projects.

I just wanted to send this as a wrap up to this phase of my work. And tell yall how infinitely appreciative I am for your feedback on this ❤️ . I suspect yall won't much like what I've ended up with, and neither do I 😆 , but at least wanted to share since this thread was so helpful to me.

😄 ❤️ 👍

**Reply by wooorm** [reply]
Glad to hear Austin! :)
